### ISSUE TITLE
Off-by-one error with Unicode?

### ISSUE THREAD

**Issue by vsajip** [javacc21/javacc21#85]
I used a logging test file from Python to test the Python lexer. In this line:
https://github.com/python/cpython/blob/f783428a2313a729ca8b539c5a86ff114b9ff375/Lib/test/test_logging.py#L4743
which is
```python
            logging.debug('😂: ☃️: The Øresund Bridge joins Copenhagen to Malmö')
```
I make the string literal as being from column 27 to column 79, but the lexer outputs 80:
```
LPAREN: ( 4743 26 4743 26
STRING_LITERAL: '😂: ☃️: The Øresund Bridge joins Copenhagen to Malmö' 4743 27 4743 80
RPAREN: ) 4743 81 4743 81
NEWLINE: 
 4743 82 4743 82
DEDENT: DEDENT 4744 1 4744 8
```
Now this might be to do with how Python and Java count Unicode chars - Python doesn't normally use surrogate pairs, for example (not that there are necessarily any surrogate pairs in the string, but you catch my drift). The positions of the `STRING_LITERAL`, `RPAREN` and `NEWLINE` are off-by-one, but the `DEDENT` is synched up and there are no other differences. Obviously not a high priority, but wanted to flag it up.

**Comment by revusky** [comment]
The problem is pretty clearly the handling of supplementary unicode characters, i.e. the ones beyond FFFF that, in fact, require a surrogate pair. Both the smiley and the snowman are supplementary characters.

Now, currently, in JavaCC21, the internal logic for locations is not adjusting for this, so is treating supplementary characters as a double-offset, which is, I guess, a bug. I must have made a mental note to deal with this later, but it is not currently handled. But what is strange about this, is that the endColumn of the string literal, it seems to me, should be 78, not 79. I copy-pasted the string literal into VSCode and it is very strange because it treats the smiley as a single character in terms of column offset, but treats the snowman as 2 characters in terms offset, though it does display both emojis. But it has the final quote as an offset of 79 when it should be 78.

The logic of JavaCC21 is that both the emojis are being treated as double-offsets, so the final offset is 80. But I think it should actually be 78. At least assuming it uses the logic code points, as opposed to code units...

**Comment by revusky** [comment]
I would add that this is a strange situation, because I'm looking at code that I wrote some months ago, and not understanding what I was thinking. For example, here: https://github.com/javacc21/javacc21/blob/master/src/ftl/java/FileLineMap.java.ftl#L250

You see there that I am very consciously (it seems...) using code unit logic, thus incrementing `column` by 2 when hitting a surrogate pair. 

Also lines 192 and 208 in the same file, you see that I increment (again) the column variable, so I seem to have been thinking consciously that column is using a *code units* logic, not *code points*. But I honestly cannot remember my thinking on this. It seems to me, looking back on this, that, in terms of reported column locations, these supplementary characters should correspond to *one* horizontal offset, not *two*. But I must have consciously thought that the latter was correct when I wrote it. 

Certainly, in terms of what a human being sees, an emoji or other supplementary character occupies one horizontal space. Internally, the 32-bit character is occupying two 16-bit character slots, of course, but....
